File: wallet_core/platform_support/ios/PlatformSupport/PlatformSupport.swift
//
//  PlatformSupport.swift
//  PlatformSupport
//
//  Created by The Wallet Developers on 13/04/2023.
//

import Foundation

public final class PlatformSupport {
    public static let shared = PlatformSupport()

    private let signingKey: SigningKey
    private let encryptionKey: EncryptionKey
    private let attestedKey: AttestedKey
    private let utilities: Utilities

    private init() {
        self.signingKey = SigningKey()
        self.encryptionKey = EncryptionKey()
        self.attestedKey = AttestedKey()
        self.utilities = Utilities()

        initPlatformSupport(
            signingKey: self.signingKey,
            encryptionKey: self.encryptionKey,
            attestedKey: self.attestedKey,
            utils: self.utilities
        )
    }
}

File: wallet_core/platform_support/android/src/main/java/nl/rijksoverheid/edi/wallet/platform_support/PlatformSupport.kt
package nl.rijksoverheid.edi.wallet.platform_support

import android.content.Context
import androidx.annotation.VisibleForTesting
import nl.rijksoverheid.edi.wallet.platform_support.attested_key.AttestedKeyBridge
import nl.rijksoverheid.edi.wallet.platform_support.keystore.encryption.EncryptionKeyBridge
import nl.rijksoverheid.edi.wallet.platform_support.keystore.signing.SigningKeyBridge
import nl.rijksoverheid.edi.wallet.platform_support.utilities.UtilitiesBridge
import nl.rijksoverheid.edi.wallet.platform_support.utilities.storage.StoragePathProviderImpl
import uniffi.platform_support.initPlatformSupport

class PlatformSupport private constructor(context: Context) {

    @VisibleForTesting
    val encryptionKeyBridge = EncryptionKeyBridge(context)

    @VisibleForTesting
    val signingKeyBridge = SigningKeyBridge(context)

    // TODO: PVW-4069: Don't hard-code google cloud project number here
    @VisibleForTesting
    val attestedKeyBridge = AttestedKeyBridge(context, 12143997365u)

    @VisibleForTesting
    val utilitiesBridge = UtilitiesBridge(StoragePathProviderImpl(context))

    init {
        initPlatformSupport(signingKeyBridge, encryptionKeyBridge, attestedKeyBridge, utilitiesBridge)
    }

    companion object {
        @Volatile
        private var INSTANCE: PlatformSupport? = null

        fun getInstance(context: Context): PlatformSupport =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: PlatformSupport(context.applicationContext).also { INSTANCE = it }
            }
    }

}
